File: javascript/0098. Validate Binary Search Tree.js
// INPUT:     5
//           / \
//          1   4
//             / \
//            3   6
// OUTPUT: false

const isValidBST = (head, minValue = -Infinity, maxValue = Infinity) => {
  if (head === null) return true;

  return (
    head.val > minValue &&
    head.val < maxValue &&
    isValidBST(head.left, minValue, head.val) &&
    isValidBST(head.right, head.val, maxValue)
  );
};

console.log(
  isValidBST({
    val: 5,
    right: {
      val: 4,
      left: {
        val: 3,
        left: null,
        right: null,
      },
      right: {
        val: 6,
        left: null,
        right: null,
      },
    },
    left: { val: 1, right: null, left: null },
  })
);

File: python/0098. Validate Binary Search Tree.py
from numbers import Number


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: TreeNode,
                   min_val: Number = float('-inf'),
                   max_val: Number = float('inf')) -> bool:
        if root is None:
            return True

        if not (min_val < root.val < max_val):
            return False

        return self.isValidBST(root.left, min_val, min(max_val, root.val)) and \
            self.isValidBST(root.right, max(min_val, root.val), max_val)
